internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,263 @@
+	autoScalerEventStart                                      = "start"
+	autoScalerEventStop                                       = "stop"
+	autoScalerEventLogMsg                     string          = "concurrency auto scaler event"
+	testTimeFormat                            string          = "15:04:05"
+)
+
+type (
+	ConcurrencyAutoScaler struct {
+		shutdownChan chan struct{}
+		wg           sync.WaitGroup
+		log          *zap.Logger
+		scope        tally.Scope
+		clock        clockwork.Clock
+
+		concurrency *ConcurrencyLimit
+		cooldown    time.Duration
+		updateTick  time.Duration
+
+		// enable auto scaler on concurrency or not
+		enable atomic.Bool
+
+		// poller
+		pollerInitCount        int
+		pollerMaxCount         int
+		pollerMinCount         int
+		pollerWaitTimeInMsLog2 *rollingAverage // log2(pollerWaitTimeInMs+1) for smoothing (ideal value is 0)
+		pollerPermitLastUpdate time.Time
+	}
+
+	ConcurrencyAutoScalerInput struct {
+		Concurrency    *ConcurrencyLimit
+		Cooldown       time.Duration // cooldown time of update
+		Tick           time.Duration // frequency of update check
+		PollerMaxCount int
+		PollerMinCount int
+		Logger         *zap.Logger
+		Scope          tally.Scope
+		Clock          clockwork.Clock
+	}
+
+	autoScalerEvent string
+)
+
+func NewConcurrencyAutoScaler(input ConcurrencyAutoScalerInput) *ConcurrencyAutoScaler {
+	tick := defaultAutoScalerUpdateTick
+	if input.Tick != 0 {
+		tick = input.Tick
+	}
+	return &ConcurrencyAutoScaler{
+		shutdownChan:           make(chan struct{}),
+		concurrency:            input.Concurrency,
+		cooldown:               input.Cooldown,
+		log:                    input.Logger,
+		scope:                  input.Scope,
+		clock:                  input.Clock,
+		updateTick:             tick,
+		enable:                 atomic.Bool{}, // initial value should be false and is only turned on from auto config hint
+		pollerInitCount:        input.Concurrency.PollerPermit.Quota(),
+		pollerMaxCount:         input.PollerMaxCount,
+		pollerMinCount:         input.PollerMinCount,
+		pollerWaitTimeInMsLog2: newRollingAverage(numberOfPollsInRollingAverage),
+		pollerPermitLastUpdate: input.Clock.Now(),
+	}
+}
+
+func (c *ConcurrencyAutoScaler) Start() {
+	c.logEvent(autoScalerEventStart)
+
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
+		ticker := c.clock.NewTicker(c.updateTick)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.shutdownChan:
+				return
+			case <-ticker.Chan():
+				c.logEvent(autoScalerEventMetrics)
+				c.updatePollerPermit()
+			}
+		}
+	}()
+}
+
+func (c *ConcurrencyAutoScaler) Stop() {
+	close(c.shutdownChan)
+	c.wg.Wait()
+	c.logEvent(autoScalerEventStop)
+}
+
+// ProcessPollerHint reads the poller response hint and take actions
+// 1. update poller wait time
+// 2. enable/disable auto scaler
+func (c *ConcurrencyAutoScaler) ProcessPollerHint(hint *shared.AutoConfigHint) {
+	if hint == nil {
+		c.log.Warn("auto config hint is nil, this results in no action")
```

Comment:
maybe just info?  otherwise this'll make a flood of warns in user services when rolling out or disabling, but it's not really an unexpected / worrying thing in that state.